### PR TITLE
Use Clock interface

### LINKED ISSUE #4562: Replace `System.currentTimeMillis` usage with injectable `Clock`
- It's good practice to avoid indeterminate behaviour in classes when unit testing

### PR DESCRIPTION
**Pull request by bmarty**
## Type of change

- [ ] Feature
- [ ] Bugfix
- [x] Technical
- [ ] Other :

## Content

- Use a `Clock` interface instead of direct usage of `System.currentTimeMillis()`
- Sometimes replace by a `UUID` or `Random.nextLong()`, when this is not strictly required to have a timestamp.
- Some formatting fix also (this is polluting a bit sadly)
- Not sure that was necessary (?) but I also replaced the usage in the tests, by implementing a `DefaultClock`.

Closes #4562 

## Motivation and context

Be able to speed up test! See #4562 

## Tests

- Quick smoke test
- Run sanity test (in progress)

## Tested devices

- [x] Physical
- [x] Emulator
- OS version(s):

## Reviewers

CC @ouchadam who opened the original issue.

### CODE REVIEW

File: matrix-sdk-android/src/androidTest/java/org/matrix/android/sdk/internal/crypto/AttachmentEncryptionTest.kt [hunk 1]
```diff
@@ -48,13 +50,18 @@ class AttachmentEncryptionTest {
         inputStream = if (inputAsByteArray.isEmpty()) {
             inputAsByteArray.inputStream()
         } else {
-            val memoryFile = MemoryFile("file" + System.currentTimeMillis(), inputAsByteArray.size)
+            val memoryFile = MemoryFile("file_" + UUID.randomUUID(), inputAsByteArray.size)
```
Comment (ouchadam) [inline matrix-sdk-android/src/androidTest/java/org/matrix/android/sdk/internal/crypto/AttachmentEncryptionTest.kt:53]: :100:

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/crypto/GossipingWorkManager.kt [hunk 2]
```diff
@@ -44,8 +45,8 @@ internal class GossipingWorkManager @Inject constructor(
     }
 
     // Prevent sending queue to stay broken after app restart
-    // The unique queue id will stay the same as long as this object is instanciated
-    val queueSuffixApp = System.currentTimeMillis()
+    // The unique queue id will stay the same as long as this object is instantiated
+    private val queueSuffixApp = UUID.randomUUID()
```
Comment (ouchadam) [inline matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/crypto/GossipingWorkManager.kt:49]: :+1:

File: tools/check/forbidden_strings_in_code.txt [hunk 3]
```diff
@@ -173,4 +173,7 @@ PreferenceManager\.getDefaultSharedPreferences==2
 # findViewById
 
 ### Do not use `template_` string. Please remove the prefix `template_` to use the generated resource instead.
-R\.string\.template_
\ No newline at end of file
+R\.string\.template_
+
+### Use the Clock interface, or use `measureTimeMillis`
+System\.currentTimeMillis\(\)===2
```
Comment (ouchadam) [inline tools/check/forbidden_strings_in_code.txt:179]: :100: nice check!

File: vector/src/main/java/im/vector/app/core/services/VectorSyncService.kt [hunk 4]
```diff
@@ -137,20 +142,27 @@ class VectorSyncService : SyncService() {
...
+
         override fun doWork(): Result {
             Timber.d("## Sync: RestartWhenNetworkOn.doWork()")
             val sessionId = inputData.getString(KEY_SESSION_ID) ?: return Result.failure()
             val syncTimeoutSeconds = inputData.getInt(KEY_SYNC_TIMEOUT_SECONDS, BackgroundSyncMode.DEFAULT_SYNC_TIMEOUT_SECONDS)
             val syncDelaySeconds = inputData.getInt(KEY_SYNC_DELAY_SECONDS, BackgroundSyncMode.DEFAULT_SYNC_DELAY_SECONDS)
             val isPeriodic = inputData.getBoolean(KEY_IS_PERIODIC, false)
+
+            // Not sure how to inject a Clock here
+            val clock = DefaultClock()
```
Comment (ouchadam) [inline vector/src/main/java/im/vector/app/core/services/VectorSyncService.kt:158]: we would need to provide this worker via factory like https://github.com/vector-im/element-android/blob/develop/matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/worker/MatrixWorkerFactory.kt#L46

I'm not sure it's worth the effort for just the clock dependency

File: vector/src/main/java/im/vector/app/features/settings/VectorPreferences.kt [hunk 5]
```diff
@@ -664,9 +668,9 @@ class VectorPreferences @Inject constructor(private val context: Context) {
      */
     fun getMinMediasLastAccessTime(): Long {
         return when (getSelectedMediasSavingPeriod()) {
-            MEDIA_SAVING_3_DAYS  -> System.currentTimeMillis() / 1000 - 3 * 24 * 60 * 60
-            MEDIA_SAVING_1_WEEK  -> System.currentTimeMillis() / 1000 - 7 * 24 * 60 * 60
-            MEDIA_SAVING_1_MONTH -> System.currentTimeMillis() / 1000 - 30 * 24 * 60 * 60
+            MEDIA_SAVING_3_DAYS  -> clock.epochMillis() / 1000 - 3 * 24 * 60 * 60
```
Comment (ouchadam) [inline vector/src/main/java/im/vector/app/features/settings/VectorPreferences.kt:671]: not for this PR, I could imagine in the future a `ClockExtensions` for these types of helpers

**Review by ouchadam — APPROVED** [review]
from a static review everything looks good to me :+1: 

would be good to double check the keybackup (as I know that area can be a little brittle~)

**Review by Florian14 — APPROVED** [review]
Just some minor comments, not blocking for the merge

**Review by ganfra — APPROVED** [review]
Just one remark, otherwise ok.

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/session/crypto/model/IncomingRequestCancellation.kt [hunk 6]
```diff
@@ -47,15 +47,15 @@ data class IncomingRequestCancellation(
          *
          * @param event the event
          */
-        fun fromEvent(event: Event): IncomingRequestCancellation? {
+        fun fromEvent(event: Event, now: Long): IncomingRequestCancellation? {
```
[the remark this answers is not in the document]
Comment (bmarty) [reply matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/session/crypto/model/IncomingRequestCancellation.kt:50]: Doc added and all `now` renamed to `currentTimeMillis`

File: vector/src/androidTest/java/im/vector/app/EspressoExt.kt [hunk 7]
```diff
@@ -168,7 +172,7 @@ fun activityIdlingResource(activityClass: Class<*>): IdlingResource {
     val res = object : IdlingResource, ActivityLifecycleCallback {
         private var callback: IdlingResource.ResourceCallback? = null
         private var resumedActivity: Activity? = null
-        private val uniqTS = System.currentTimeMillis()
+        private val uniqTS = Random.nextLong()
```
[the remark this answers is not in the document]
Comment (bmarty) [reply vector/src/androidTest/java/im/vector/app/EspressoExt.kt:175]: I inlined it :)

File: vector/src/main/java/im/vector/app/features/home/room/detail/composer/MessageComposerViewState.kt [hunk 8]
```diff
@@ -35,7 +36,7 @@ sealed interface SendMode {
             val text: String,
             val fromSharing: Boolean,
             // This is necessary for forcing refresh on selectSubscribe
-            private val ts: Long = System.currentTimeMillis()
+            private val ts: Int = Random.nextInt()
```
[the remark this answers is not in the document]
Comment (bmarty) [reply vector/src/main/java/im/vector/app/features/home/room/detail/composer/MessageComposerViewState.kt:39]: renamed

**Review by Florian14 — APPROVED** [review]
SGTM, thanks

**Pull request merged**